File: src/day13/day13.py
from functools import cmp_to_key

def main():
    print(part1())
    print(part2())

def cmp(l1, l2):

    if type(l1) == int and type(l2) == list:
        l1 = [l1]
    if type(l1) == list and type(l2) == int:
        l2 = [l2]
        
    if type(l1) == int and type(l2) == int:
        
        if l1 < l2:
            return 1
        if l1 == l2:
            return 0

        return -1
    
    else:

        for i in range(max(len(l1), len(l2))):

            if i == len(l1):
                if len(l1) == len(l2):
                    return 0
                return 1
            
            if i == len(l2):
                return -1

            x = cmp(l1[i], l2[i])

            if x == 0:
                continue
            else:
                return x

    return 0

def part1():
    
    res = 0
    pi = 1

    with open("src/day13/input.txt", "r") as f:

        for line in f:

            if line == "\n":
                continue

            l1 = eval(line)

            nextline = next(f, None)

            l2 = eval(nextline)

            x = cmp(l1, l2)

            if x == 1:

                res += pi

            pi += 1


    return res

def part2():

    s = []

    with open("src/day13/input.txt", "r") as f:

        for line in f:

            if line == "\n":
                continue

            l1 = eval(line)

            nextline = next(f, None)

            l2 = eval(nextline)

            s.append(l1)
            s.append(l2)

    sorted_s = sorted(s, key=cmp_to_key(cmp), reverse=True)

    return (sorted_s.index([[2]]) + 1) * (sorted_s.index([[6]]) + 1)

if __name__ == "__main__":
    main()

File: src/day13/day13.rs
// use std::{cmp, fs};

// struct List {

//     children: Option<Vec<Box<List>>>, 

// }

// pub fn run() -> (u32, u32) {

//     let input = fs::read_to_string("src/day13/test.txt")
//     .expect("Should have been able to read the file");

//     let p1 = part1(&input);
//     let p2 = part2(&input);

//     return (p1, p2);

// }

// fn parse_list(list: &str) -> &Box<List> {

//     let mut cur_list = Box::new(List {children: None});
//     let head = &cur_list;



//     return head;

// }

// fn part1(input: &str) -> u32 {

//     let mut lines = input.lines();

//     while let (Some(a), Some(b), Some(_)) = (lines.next(), lines.next(), lines.next()) {

//         println!("{a}");
//         println!("{b}");

//         parse_list(a);

//         break;

//     }

//     0

// }

// fn part2(input: &str) -> u32 {0}